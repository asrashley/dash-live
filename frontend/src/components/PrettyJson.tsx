import { Fragment } from "preact/jsx-runtime";

interface ValueProps {
    val: any;
}

function Value({ val }: ValueProps) {
  if (typeof val === "string") {
    return <span className="value">{`"${val}"`}</span>;
  }
  if (Array.isArray(val)) {
    return <Fragment>{`[`}{val.map((item) => <Value val={item} />)}{`]`}</Fragment>;
  }
  if (typeof val === "object") {
    return <PrettyJson data={val} />;
  }
  const sval = `${val}`;
  return <span className="value">{sval}</span>;
}

interface KeyValueProps {
    keyName: string;
    value: any;
    idx: number;
}
function KeyValue({ keyName, value, idx }: KeyValueProps) {
  return <Fragment>
    {idx > 0 ? ", " : ""}<span className="key">{keyName}</span>: <Value val={value} />
    </Fragment>;
}

export interface PrettyJsonProps {
  className?: string;
  data: object;
}

export function PrettyJson({ data = {}, className="" }: PrettyJsonProps) {
  const cn = `json bg-secondary-subtle border border-dark-subtle ${className}`;
  return (
    <div className={cn}>
      {Object.entries(data).map(([key, value], idx) => (
        <KeyValue key={key} idx={idx} keyName={key} value={value} />
      ))}
    </div>
  );
}
