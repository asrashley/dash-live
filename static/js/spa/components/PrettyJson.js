import { html } from "htm/preact";

function Value({val}) {
    if (typeof val === "string") {
        return html`<span className="value">"${val}"</span>`;
    }
    if (Array.isArray(val)) {
        return html`[${val.map((item) => html`<${Value} val=${item}><//>`)}]`;

    }
    if (typeof val === "object") {
        return html`<${PrettyJson} data=${val} />`;
    }
    const sval = `${val}`;
    return html`<span className="value">${sval}</span>`;
}

function KeyValue({keyName, value, idx}) {
    return html`${idx > 0 ? ', ': ''}<span className="key">${keyName}</span>: <${Value} val=${value} />`;
}

export function PrettyJson({data = {}, className}) {
    return html`<div className=${`json bg-secondary-subtle border border-dark-subtle ${className}`}>{
    ${ Object.entries(data).map(([key, value], idx) =>
        html`<${KeyValue} key=${key} idx=${idx} keyName=${key} value=${value} />`)}
}</div>`;
}
