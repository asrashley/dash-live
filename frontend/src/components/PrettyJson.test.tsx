import { describe, expect, test } from "vitest";

import { renderWithProviders } from "../test/renderWithProviders";
import { PrettyJson } from "./PrettyJson";

describe("PrettyJson", () => {
  const data = {
    hello: "world",
    number: 32,
    aBool: true,
    anArray: [
      "apple",
      "pear",
      {
        strawberry: "jam",
      },
    ],
    anObject: {
        james: 'bond',
    },
  };

  test("should match snapshot", () => {
    const { asFragment } = renderWithProviders(
      <PrettyJson className="pretty" data={data} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
