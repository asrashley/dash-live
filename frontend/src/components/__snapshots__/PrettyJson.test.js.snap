// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrettyJson > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="json bg-secondary-subtle border border-dark-subtle pretty"
  >
    {
    <span
      class="key"
    >
      hello
    </span>
    : 
    <span
      class="value"
    >
      "world"
    </span>
    , 
    <span
      class="key"
    >
      number
    </span>
    : 
    <span
      class="value"
    >
      32
    </span>
    , 
    <span
      class="key"
    >
      aBool
    </span>
    : 
    <span
      class="value"
    >
      true
    </span>
    , 
    <span
      class="key"
    >
      anArray
    </span>
    : [
    <span
      class="value"
    >
      "apple"
    </span>
    <span
      class="value"
    >
      "pear"
    </span>
    <div
      class="json bg-secondary-subtle border border-dark-subtle undefined"
    >
      {
      <span
        class="key"
      >
        strawberry
      </span>
      : 
      <span
        class="value"
      >
        "jam"
      </span>
      }
    </div>
    ], 
    <span
      class="key"
    >
      anObject
    </span>
    : 
    <div
      class="json bg-secondary-subtle border border-dark-subtle undefined"
    >
      {
      <span
        class="key"
      >
        james
      </span>
      : 
      <span
        class="value"
      >
        "bond"
      </span>
      }
    </div>
    }
  </div>
</DocumentFragment>
`;
